otherlibs/systhreads/st_stubs.c

```diff
@@ -532,9 +532,9 @@ static ST_THREAD_FUNCTION caml_thread_start(void * arg)
 
   /* Associate the thread descriptor with the thread */
   st_tls_set(thread_descriptor_key, (void *) th);
-  st_thread_set_id(Ident(th->descr));
   /* Acquire the global mutex */
   caml_leave_blocking_section();
+  st_thread_set_id(Ident(th->descr));
```

Comment:
If anyone is wondering about what's wrong in the original code: `th->descr` points to a block allocated in the OCaml heap.  So, if we don't hold the master lock, the block may be in the process of being moved by the GC or the compactor, and accessing it is a race condition.

Reply:
Thanks, added the explanation to the commit message.